cloud/pkg/cloudhub/handler/eventhandler.go

```diff
@@ -40,10 +43,93 @@ type EventHandle struct {
 	EventQueue        *channelq.ChannelEventQueue
 	Context           *context.Context
 	Handlers          []HandleFunc
+	NodeLimit         int
+	KeepaliveChannel  map[string]chan struct{}
 }
 
 type HandleFunc func(hi hubio.CloudHubIO, info *emodel.HubInfo, stop chan ExitCode)
 
+var once sync.Once
+
+// CloudhubHandler the shared handler for both websocket and quic servers
+var CloudhubHandler *EventHandle
+
+// InitHandler create a handler for websocket and quic servers
+func InitHandler(config *util.Config, eventq *channelq.ChannelEventQueue, c *context.Context) {
+	once.Do(func() {
+		CloudhubHandler = &EventHandle{
+			KeepaliveInterval: config.KeepaliveInterval,
+			WriteTimeout:      config.WriteTimeout,
+			EventQueue:        eventq,
+			NodeLimit:         config.NodeLimit,
+			Context:           c,
+		}
+
+		CloudhubHandler.KeepaliveChannel = make(map[string]chan struct{})
+		CloudhubHandler.Handlers = []HandleFunc{CloudhubHandler.KeepaliveCheckLoop, CloudhubHandler.EventWriteLoop}
+
+		CloudhubHandler.initServerEntries()
+	})
+}
+
+// initServerEntries regist handler func
+func (eh *EventHandle) initServerEntries() {
+	mux.Entry(mux.NewPattern("*").Op("*"), eh.HandleServer)
+}
+
+// HandleServer handle all the request from node
+func (eh *EventHandle) HandleServer(container *mux.MessageContainer, writer mux.ResponseWriter) {
+	nodeID := container.Header.Get("node_id")
+	projectID := container.Header.Get("project_id")
+
+	if eh.GetNodeCount() >= eh.NodeLimit {
+		bhLog.LOGGER.Errorf("Fail to serve node %s, reach node limit", nodeID)
+		return
+	}
+
+	if container.Message.GetOperation() == emodel.OpKeepalive {
+		bhLog.LOGGER.Infof("Keepalive message received from node: %s", nodeID)
+		eh.KeepaliveChannel[nodeID] <- struct{}{}
+		return
+	}
+
+	err := eh.Pub2Controller(&emodel.HubInfo{ProjectID: projectID, NodeID: nodeID}, container.Message)
+	if err != nil {
+		// if err, we should stop node, write data to edgehub, stop nodify
+		bhLog.LOGGER.Errorf("Failed to serve handle with error: %s", err.Error())
+	}
+}
+
+// OnRegister regist node on first connection
+func (eh *EventHandle) OnRegister(connection conn.Connection) {
+	nodeID := connection.ConnectionState().Headers.Get("node_id")
+	projectID := connection.ConnectionState().Headers.Get("project_id")
+
+	if _, ok := eh.KeepaliveChannel[nodeID]; !ok {
+		eh.KeepaliveChannel[nodeID] = make(chan struct{}, 1)
+	}
+
+	io := &hubio.JSONIO{Connection: connection}
+	go eh.ServeConn(io, &emodel.HubInfo{ProjectID: projectID, NodeID: nodeID})
+}
+
+// KeepaliveCheckLoop checks whether the edge node is still alive
+func (eh *EventHandle) KeepaliveCheckLoop(hi hubio.CloudHubIO, info *emodel.HubInfo, stop chan ExitCode) {
+	for {
+		keepaliveTimer := time.NewTimer(time.Duration(eh.KeepaliveInterval) * time.Second)
+		select {
+		case <-eh.KeepaliveChannel[info.NodeID]:
+			bhLog.LOGGER.Infof("Node %s is still alive", info.NodeID)
+			keepaliveTimer.Stop()
```

Comment:
@shouhong , can you please check ?

Reply:
@rohitsardesai83 @kexun The logic here is after receiving a keepalive message from a node, stop the current timer, then go to the next loop to start a new timeout timer and wait for the the next keepalive message. If we add return here, it will exit the whole function instead of going to the next loop. 
So we cannot return after keepaliveTimer.Stop().

Reply:
@shouhong you are right.